File: src/discover.ts
import {existsSync, readdirSync, realpathSync, statSync} from "node:fs";
import path from "node:path";
import {getTopLevel, getWorktreeList, parseWorktreePaths} from "./git.ts";

export type RepoSource = "scan" | "worktree";

export type DiscoveredRepo = {
  path: string;
  realPath: string;
  displayName: string;
  source: RepoSource;
};

export type DiscoverOptions = {
  cwd: string;
  maxDepth: number;
  verbose?: boolean;
  warn?: (message: string) => void;
};

export const IGNORED_DIR_NAMES = new Set([
  "node_modules",
  "dist",
  "build",
  "public",
  ".gradle",
  ".idea",
  ".vscode",
  "target",
  "coverage",
  ".next",
  ".nuxt",
  ".cache",
  ".terraform",
  ".turbo",
  ".parcel-cache",
  "vendor",
  "out",
  "tmp",
  "temp",
  "__pycache__",
  ".venv",
  "venv",
  ".mypy_cache",
  ".pytest_cache",
  ".tox",
  ".yarn",
  ".pnpm-store",
  ".svelte-kit",
  ".angular",
  ".serverless",
  ".wrangler",
  ".netlify",
  ".vercel",
  ".expo",
  ".docusaurus",
  ".storybook-static",
  ".astro",
  ".remix",
  ".output",
  ".cache-loader",
  ".rustup",
  ".cargo",
  "Pods",
  "DerivedData",
  "bin",
  "obj",
  "logs",
  "log",
]);

export function findGitCandidates(options: {
  cwd: string;
  maxDepth: number;
  ignoredDirNames?: ReadonlySet<string>;
}): string[] {
  const cwd = path.resolve(options.cwd);
  const ignoredDirNames = options.ignoredDirNames ?? IGNORED_DIR_NAMES;
  const candidates = new Set<string>();

  function walk(directory: string, depth: number): void {
    let entries;
    try {
      entries = readdirSync(directory, {withFileTypes: true});
    } catch {
      return;
    }

    for (const entry of entries) {
      const entryPath = path.join(directory, entry.name);

      if (entry.name === ".git") {
        if (depth > 0 && depth <= options.maxDepth && (entry.isDirectory() || entry.isFile())) {
          candidates.add(directory);
        }
        continue;
      }

      if (!entry.isDirectory() || entry.isSymbolicLink()) {
        continue;
      }

      if (ignoredDirNames.has(entry.name)) {
        continue;
      }

      if (depth < options.maxDepth) {
        walk(entryPath, depth + 1);
      }
    }
  }

  walk(cwd, 0);

  return [...candidates].sort((left, right) => displayNameForPath(cwd, left).localeCompare(displayNameForPath(cwd, right)));
}

export function discoverRepos(options: DiscoverOptions): DiscoveredRepo[] {
  const cwd = path.resolve(options.cwd);
  const reposByRealPath = new Map<string, DiscoveredRepo>();
  const warn = createWarner(options);

  for (const candidate of findGitCandidates({cwd, maxDepth: options.maxDepth})) {
    const verified = verifyRepo(candidate, cwd, "scan", warn);
    if (verified !== undefined) {
      reposByRealPath.set(verified.realPath, verified);
    }
  }

  const scannedRepos = [...reposByRealPath.values()];
  for (const repo of scannedRepos) {
    const result = getWorktreeList(repo.path);
    if (!result.ok) {
      warn(`Failed to list worktrees for ${repo.displayName}: ${result.stderr.trim() || `git exited ${result.status ?? "unknown"}`}`);
      continue;
    }

    for (const worktreePath of parseWorktreePaths(result.stdout)) {
      const verified = verifyRepo(worktreePath, cwd, "worktree", warn);
      if (verified !== undefined && !reposByRealPath.has(verified.realPath)) {
        reposByRealPath.set(verified.realPath, verified);
      }
    }
  }

  return [...reposByRealPath.values()].sort((left, right) => left.displayName.localeCompare(right.displayName));
}

function verifyRepo(repoPath: string, cwd: string, source: RepoSource, warn: (message: string) => void): DiscoveredRepo | undefined {
  if (!existsSync(repoPath)) {
    warn(`Skipping missing repo path: ${repoPath}`);
    return undefined;
  }

  let repoRealPath: string;
  try {
    const stats = statSync(repoPath);
    if (!stats.isDirectory()) {
      warn(`Skipping non-directory repo path: ${repoPath}`);
      return undefined;
    }
    repoRealPath = realpathSync(repoPath);
  } catch (error) {
    warn(`Skipping inaccessible repo path ${repoPath}: ${formatError(error)}`);
    return undefined;
  }

  const topLevel = getTopLevel(repoPath);
  if (!topLevel.ok) {
    warn(`Skipping invalid git repo ${displayNameForPath(cwd, repoPath)}: ${topLevel.stderr.trim() || `git exited ${topLevel.status ?? "unknown"}`}`);
    return undefined;
  }

  const topLevelPath = topLevel.stdout.trim();
  let topLevelRealPath: string;
  try {
    topLevelRealPath = realpathSync(topLevelPath);
  } catch (error) {
    warn(`Skipping repo ${displayNameForPath(cwd, repoPath)} with inaccessible top-level ${topLevelPath}: ${formatError(error)}`);
    return undefined;
  }

  if (repoRealPath !== topLevelRealPath) {
    warn(`Skipping nested git directory ${displayNameForPath(cwd, repoPath)}; top-level is ${topLevelPath}`);
    return undefined;
  }

  return {
    path: repoPath,
    realPath: repoRealPath,
    displayName: displayNameForPath(cwd, repoPath),
    source,
  };
}

export function displayNameForPath(cwd: string, repoPath: string): string {
  const relativePath = path.relative(path.resolve(cwd), path.resolve(repoPath));

  if (relativePath === "") {
    return ".";
  }

  if (!relativePath.startsWith("..") && !path.isAbsolute(relativePath)) {
    return relativePath;
  }

  return path.resolve(repoPath);
}

function createWarner(options: DiscoverOptions): (message: string) => void {
  return (message: string) => {
    if (options.verbose !== true) {
      return;
    }
    if (options.warn !== undefined) {
      options.warn(message);
      return;
    }
    process.stderr.write(`gitsy: warning: ${message}\n`);
  };
}

function formatError(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}

File: internal/discover/discover.go
package discover

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/flexdinesh/gitsy/internal/git"
)

type RepoSource string

const (
	SourceScan     RepoSource = "scan"
	SourceWorktree RepoSource = "worktree"
)

type Repo struct {
	Path        string
	RealPath    string
	DisplayName string
	Source      RepoSource
}

type Options struct {
	Cwd      string
	MaxDepth int
	Verbose  bool
	Warn     func(message string)
}

var IgnoredDirNames = map[string]struct{}{
	"node_modules":      {},
	"dist":              {},
	"build":             {},
	"public":            {},
	".gradle":           {},
	".idea":             {},
	".vscode":           {},
	"target":            {},
	"coverage":          {},
	".next":             {},
	".nuxt":             {},
	".cache":            {},
	".terraform":        {},
	".turbo":            {},
	".parcel-cache":     {},
	"vendor":            {},
	"out":               {},
	"tmp":               {},
	"temp":              {},
	"__pycache__":       {},
	".venv":             {},
	"venv":              {},
	".mypy_cache":       {},
	".pytest_cache":     {},
	".tox":              {},
	".yarn":             {},
	".pnpm-store":       {},
	".svelte-kit":       {},
	".angular":          {},
	".serverless":       {},
	".wrangler":         {},
	".netlify":          {},
	".vercel":           {},
	".expo":             {},
	".docusaurus":       {},
	".storybook-static": {},
	".astro":            {},
	".remix":            {},
	".output":           {},
	".cache-loader":     {},
	".rustup":           {},
	".cargo":            {},
	"Pods":              {},
	"DerivedData":       {},
	"bin":               {},
	"obj":               {},
	"logs":              {},
	"log":               {},
}

func FindGitCandidates(cwd string, maxDepth int, ignoredDirNames map[string]struct{}) []string {
	root, err := filepath.Abs(cwd)
	if err != nil {
		root = filepath.Clean(cwd)
	}
	if ignoredDirNames == nil {
		ignoredDirNames = IgnoredDirNames
	}

	candidates := map[string]struct{}{}
	var walk func(directory string, depth int)
	walk = func(directory string, depth int) {
		entries, err := os.ReadDir(directory)
		if err != nil {
			return
		}

		for _, entry := range entries {
			entryPath := filepath.Join(directory, entry.Name())

			if entry.Name() == ".git" {
				if depth > 0 && depth <= maxDepth && isGitMarker(entryPath) {
					candidates[directory] = struct{}{}
				}
				continue
			}

			if !entry.IsDir() || entry.Type()&os.ModeSymlink != 0 {
				continue
			}
			if _, ignored := ignoredDirNames[entry.Name()]; ignored {
				continue
			}
			if depth < maxDepth {
				walk(entryPath, depth+1)
			}
		}
	}

	walk(root, 0)

	result := make([]string, 0, len(candidates))
	for candidate := range candidates {
		result = append(result, candidate)
	}
	sort.Slice(result, func(i, j int) bool {
		return DisplayNameForPath(root, result[i]) < DisplayNameForPath(root, result[j])
	})
	return result
}

func Discover(options Options) []Repo {
	return DiscoverContext(context.Background(), options)
}

func DiscoverContext(ctx context.Context, options Options) []Repo {
	cwd, err := filepath.Abs(options.Cwd)
	if err != nil {
		cwd = filepath.Clean(options.Cwd)
	}

	warn := createWarner(options)
	reposByRealPath := map[string]Repo{}

	for _, candidate := range FindGitCandidates(cwd, options.MaxDepth, nil) {
		verified, ok := verifyRepo(ctx, candidate, cwd, SourceScan, warn)
		if ok {
			reposByRealPath[verified.RealPath] = verified
		}
	}

	scannedRepos := make([]Repo, 0, len(reposByRealPath))
	for _, repo := range reposByRealPath {
		scannedRepos = append(scannedRepos, repo)
	}

	for _, repo := range scannedRepos {
		result := git.WorktreeListContext(ctx, repo.Path)
		if !result.OK {
			warn(fmt.Sprintf("Failed to list worktrees for %s: %s", repo.DisplayName, gitError(result)))
			continue
		}

		for _, worktreePath := range git.ParseWorktreePaths(result.Stdout) {
			verified, ok := verifyRepo(ctx, worktreePath, cwd, SourceWorktree, warn)
			if ok {
				if _, exists := reposByRealPath[verified.RealPath]; !exists {
					reposByRealPath[verified.RealPath] = verified
				}
			}
		}
	}

	repos := make([]Repo, 0, len(reposByRealPath))
	for _, repo := range reposByRealPath {
		repos = append(repos, repo)
	}
	sort.Slice(repos, func(i, j int) bool {
		return repos[i].DisplayName < repos[j].DisplayName
	})
	return repos
}

func DisplayNameForPath(cwd string, repoPath string) string {
	root, err := filepath.Abs(cwd)
	if err != nil {
		root = filepath.Clean(cwd)
	}
	absoluteRepo, err := filepath.Abs(repoPath)
	if err != nil {
		absoluteRepo = filepath.Clean(repoPath)
	}

	relative, err := filepath.Rel(root, absoluteRepo)
	if err == nil {
		if relative == "." {
			return "."
		}
		if !strings.HasPrefix(relative, ".."+string(filepath.Separator)) && relative != ".." && !filepath.IsAbs(relative) {
			return relative
		}
	}
	return absoluteRepo
}

func verifyRepo(ctx context.Context, repoPath string, cwd string, source RepoSource, warn func(message string)) (Repo, bool) {
	stats, err := os.Stat(repoPath)
	if err != nil {
		if os.IsNotExist(err) {
			warn(fmt.Sprintf("Skipping missing repo path: %s", repoPath))
		} else {
			warn(fmt.Sprintf("Skipping inaccessible repo path %s: %s", repoPath, err.Error()))
		}
		return Repo{}, false
	}
	if !stats.IsDir() {
		warn(fmt.Sprintf("Skipping non-directory repo path: %s", repoPath))
		return Repo{}, false
	}

	repoRealPath, err := filepath.EvalSymlinks(repoPath)
	if err != nil {
		warn(fmt.Sprintf("Skipping inaccessible repo path %s: %s", repoPath, err.Error()))
		return Repo{}, false
	}

	topLevel := git.TopLevelContext(ctx, repoPath)
	if !topLevel.OK {
		warn(fmt.Sprintf("Skipping invalid git repo %s: %s", DisplayNameForPath(cwd, repoPath), gitError(topLevel)))
		return Repo{}, false
	}

	topLevelPath := strings.TrimSpace(topLevel.Stdout)
	topLevelRealPath, err := filepath.EvalSymlinks(topLevelPath)
	if err != nil {
		warn(fmt.Sprintf("Skipping repo %s with inaccessible top-level %s: %s", DisplayNameForPath(cwd, repoPath), topLevelPath, err.Error()))
		return Repo{}, false
	}

	if filepath.Clean(repoRealPath) != filepath.Clean(topLevelRealPath) {
		warn(fmt.Sprintf("Skipping nested git directory %s; top-level is %s", DisplayNameForPath(cwd, repoPath), topLevelPath))
		return Repo{}, false
	}

	return Repo{
		Path:        repoPath,
		RealPath:    repoRealPath,
		DisplayName: DisplayNameForPath(cwd, repoPath),
		Source:      source,
	}, true
}

func isGitMarker(path string) bool {
	stats, err := os.Stat(path)
	if err != nil {
		return false
	}
	return stats.IsDir() || stats.Mode().IsRegular()
}

func createWarner(options Options) func(message string) {
	return func(message string) {
		if !options.Verbose {
			return
		}
		if options.Warn != nil {
			options.Warn(message)
		}
	}
}

func gitError(result git.Result) string {
	stderr := strings.TrimSpace(result.Stderr)
	if stderr != "" {
		return stderr
	}
	return fmt.Sprintf("git exited %d", result.Status)
}
